pkg/noderesourcetopology/least_numa.go

```diff
@@ -0,0 +1,192 @@
+package noderesourcetopology
+
+import (
+	"fmt"
+
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+	"gonum.org/v1/gonum/stat/combin"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
+)
+
+func leastNUMAContainerScopeScore(pod *v1.Pod, zones topologyv1alpha1.ZoneList) (int64, *framework.Status) {
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+
+	maxNuma := 0
+	// the order how TopologyManager asks for hint is important so doing it in the same order
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/topologymanager/scope_container.go#L52
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		identifier := fmt.Sprintf("%s/%s/%s", pod.Namespace, pod.Name, container.Name)
+		numaNodes, res := numaNodesRequired(identifier, qos, nodes, container.Resources.Requests)
+		if res < 1 {
+			errMsg := fmt.Sprintf("cannot calculate NUMA score for container: %s", container.Name)
+			klog.ErrorS(fmt.Errorf(errMsg), "failed to calculate numaNodesRequired for: %s", identifier)
+			return 0, framework.NewStatus(framework.Error, errMsg)
+		}
+
+		if res > maxNuma {
+			maxNuma = res
+		}
+
+		// subtract the resources requested by the container from the given NUMA.
+		// this is necessary, so we won't allocate the same resources for the upcoming containers
+		subtractFromNUMAs(nodes, numaNodes.GetBits(), container.Resources.Requests)
+	}
+
+	return normalizeScore(maxNuma), nil
+}
+
+func leastNUMAPodScopeScore(pod *v1.Pod, zones topologyv1alpha1.ZoneList) (int64, *framework.Status) {
+	nodes := createNUMANodeList(zones)
+	qos := v1qos.GetPodQOS(pod)
+
+	identifier := fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)
+
+	resources := util.GetPodEffectiveRequest(pod)
+
+	_, res := numaNodesRequired(identifier, qos, nodes, resources)
+	if res < 0 {
+		errMsg := fmt.Sprintf("cannot calculate NUMA score for container: %s", identifier)
+		klog.Error(errMsg)
+		return 0, framework.NewStatus(framework.Error, errMsg)
+	}
+
+	return normalizeScore(res), nil
+}
+
+func normalizeScore(numaNodes int) int64 {
+	numaScore := framework.MaxNodeScore / highestNUMAID
+	return framework.MaxNodeScore - int64(numaNodes)*numaScore
+}
+
+func subtractFromNUMAs(numaNodes NUMANodeList, nodes []int, resources v1.ResourceList) {
```

Comment:
any chance to reuse/refactor `subtractFromNUMA` to fit here?

Reply:
will do, I have started this implemetation as separate plugin and moved this inside `TopologyMatch` later, so there can be more places which can reuse/refeactor functions from this package.

Reply:
I have remove old `subtractFromNUMA` function and make `filter` plugin to use this new `subtractFromNUMAs` function.

Reply:
ok, this means I need to spend a bit more time checking the two functions, then.

Reply:
thanks, I'm still forming thoughts

Reply:
after some thinking, I'm more confident on the proposed approach. I think is good enough for now. I'm still not completely happy with how the code looks, but this alone is not anymore sufficient to hold this PR. I'll have a final pass and if everything else looks good (as IIRC it should) I'll LGTM, and we can iterate over improvements on future PR.

Reply:
ok, I'll make `subtractFromNodes` a function of `NUMANodeList` in this PR. Let me just implement this today 😄

Reply:
changed

Reply:
I'm not convinced this is the best way forward, but I'll surely evaluate. Thanks for addressing the comments quickly!

Reply:
@ffromani sorry for the confusion i haven't noticed your git diff for the function and I assumed the `proposed` approach is the struct func 🤦  I like the variadic function approach better, it gets rid of passing `[]int{idx}` as param. Let me change this PR quickly.

Reply:
I have changed the implementation to apply you patch 😄

Reply:
no problem!